File: TS/系统/05．Buff系统/06．负面效果免疫状态.ts
/** @noSelfInFile */

const jass = require("jass.common") as any;
const buffTableMod = require("系统.05．Buff系统.01．Buff表") as {
  buffs: Record<string, { type?: string } | undefined>;
};
const { addDelayedCallback, getServerTime } = require("系统.00．核心系统.05．中心计时器") as {
  addDelayedCallback: (this: void, delayMs: number, callback: (this: void) => void) => number;
  getServerTime: (this: void) => number;
};

const GetHandleId = jass.GetHandleId as (handle: any) => number;
const IsUnitType = jass.IsUnitType as (unit: any, unitType: any) => boolean;
const UnitAddType = jass.UnitAddType as (unit: any, unitType: any) => boolean;
const UnitRemoveType = jass.UnitRemoveType as (unit: any, unitType: any) => boolean;
const UNIT_TYPE_SAPPER = jass.UNIT_TYPE_SAPPER;

export type 负面效果免疫范围 = "全部负面" | "控制" | "魔法负面";

export interface 施加单位负面效果免疫参数 {
  /** 默认“全部负面”。控制是负面效果的子集，因此“全部负面”会包含控制。 */
  范围?: 负面效果免疫范围 | 负面效果免疫范围[];
  /** 默认 true。用自爆工兵类别配合物编 targetsAllowed=nonsapper 拦截原生技能目标。 */
  同步原生技能目标免疫?: boolean;
}

interface 负面效果免疫行 {
  单位: any;
  原本是自爆工兵: boolean;
  全部负面免疫到期: number;
  控制负面免疫到期: number;
  魔法负面免疫到期: number;
}

const 负面效果免疫表: Record<number, 负面效果免疫行 | undefined> = {};

function 取单位句柄(单位: any): number {
  if (单位 == null || 单位 === 0) return 0;
  return GetHandleId(单位) || 0;
}

function 类型匹配(typeName: string | undefined, typePrefix: string): boolean {
  if (typeof typeName !== "string" || typeName === "" || typePrefix === "") return false;
  return typeName.substring(0, typePrefix.length) === typePrefix;
}

function 范围包含(配置范围: 负面效果免疫范围 | 负面效果免疫范围[] | undefined, 目标范围: 负面效果免疫范围): boolean {
  if (配置范围 == null) return 目标范围 === "全部负面";
  if (typeof 配置范围 === "string") return 配置范围 === 目标范围;
  for (let i = 0; i < 配置范围.length; i++) {
    if (配置范围[i] === 目标范围) return true;
  }
  return false;
}

function 取或建负面效果免疫行(单位: any): 负面效果免疫行 | undefined {
  const hid = 取单位句柄(单位);
  if (hid === 0) return undefined;
  let row = 负面效果免疫表[hid];
  if (row == null) {
    row = {
      单位,
      原本是自爆工兵: IsUnitType(单位, UNIT_TYPE_SAPPER) === true,
      全部负面免疫到期: 0,
      控制负面免疫到期: 0,
      魔法负面免疫到期: 0,
    };
    负面效果免疫表[hid] = row;
  } else {
    row.单位 = 单位;
  }
  return row;
}

function 行是否仍有负面效果免疫(row: 负面效果免疫行, now: number): boolean {
  return row.全部负面免疫到期 > now || row.控制负面免疫到期 > now || row.魔法负面免疫到期 > now;
}

function 尝试清理负面效果免疫行(this: void, 单位: any): void {
  const hid = 取单位句柄(单位);
  if (hid === 0) return;
  const row = 负面效果免疫表[hid];
  if (row == null) return;

  const now = getServerTime();
  if (行是否仍有负面效果免疫(row, now)) return;

  if (!row.原本是自爆工兵) {
    UnitRemoveType(row.单位, UNIT_TYPE_SAPPER);
  }
  delete 负面效果免疫表[hid];
}

function on负面效果免疫到期(this: void): void {
  const now = getServerTime();
  for (const key in 负面效果免疫表) {
    const hid = key as any as number;
    const row = 负面效果免疫表[hid];
    if (row != null && !行是否仍有负面效果免疫(row, now)) {
      if (!row.原本是自爆工兵) {
        UnitRemoveType(row.单位, UNIT_TYPE_SAPPER);
      }
      delete 负面效果免疫表[hid];
    }
  }
}

export function 施加单位负面效果免疫(单位: any, 持续时间: number, 参数: 施加单位负面效果免疫参数 = {}): void {
  if (单位 == null || 单位 === 0 || !(持续时间 > 0)) return;
  const row = 取或建负面效果免疫行(单位);
  if (row == null) return;

  const now = getServerTime();
  const until = now + 持续时间 * 1000;
  const 范围 = 参数.范围;

  if (范围包含(范围, "全部负面")) row.全部负面免疫到期 = row.全部负面免疫到期 > until ? row.全部负面免疫到期 : until;
  if (范围包含(范围, "控制")) row.控制负面免疫到期 = row.控制负面免疫到期 > until ? row.控制负面免疫到期 : until;
  if (范围包含(范围, "魔法负面")) row.魔法负面免疫到期 = row.魔法负面免疫到期 > until ? row.魔法负面免疫到期 : until;

  if (参数.同步原生技能目标免疫 !== false) {
    UnitAddType(单位, UNIT_TYPE_SAPPER);
  }

  addDelayedCallback(持续时间 * 1000 + 50, on负面效果免疫到期);
}

export function 施加单位控制负面效果免疫(单位: any, 持续时间: number, 同步原生技能目标免疫: boolean = true): void {
  施加单位负面效果免疫(单位, 持续时间, { 范围: "控制", 同步原生技能目标免疫 });
}

export function 施加单位魔法负面效果免疫(单位: any, 持续时间: number, 同步原生技能目标免疫: boolean = true): void {
  施加单位负面效果免疫(单位, 持续时间, { 范围: "魔法负面", 同步原生技能目标免疫 });
}

export function 清除单位负面效果免疫(单位: any): void {
  const hid = 取单位句柄(单位);
  if (hid === 0) return;
  const row = 负面效果免疫表[hid];
  if (row == null) return;
  row.全部负面免疫到期 = 0;
  row.控制负面免疫到期 = 0;
  row.魔法负面免疫到期 = 0;
  尝试清理负面效果免疫行(单位);
}

export function 单位是否免疫负面效果类型(单位: any, typeName: string | undefined): boolean {
  const hid = 取单位句柄(单位);
  if (hid === 0) return false;
  const row = 负面效果免疫表[hid];
  if (row == null) return false;

  const now = getServerTime();
  if (!行是否仍有负面效果免疫(row, now)) {
    尝试清理负面效果免疫行(单位);
    return false;
  }

  if (row.全部负面免疫到期 > now && 类型匹配(typeName, "Debuff:")) return true;
  if (row.控制负面免疫到期 > now && 类型匹配(typeName, "Debuff:control")) return true;
  if (row.魔法负面免疫到期 > now && 类型匹配(typeName, "Debuff:magic")) return true;
  return false;
}

export function 单位是否免疫负面效果BuffID(单位: any, buffID: string): boolean {
  const meta = buffTableMod.buffs[buffID];
  return 单位是否免疫负面效果类型(单位, meta?.type);
}

export {};

File: src/系统/05．Buff系统/06．负面效果免疫状态.lua
local ____lualib = require("lualib_bundle")
local __TS__StringSubstring = ____lualib.__TS__StringSubstring
local __TS__Delete = ____lualib.__TS__Delete
local ____exports = {}
---
-- @noSelfInFile
local jass = require("jass.common")
local buffTableMod = require("系统.05．Buff系统.01．Buff表")
local ____require_result_0 = require("系统.00．核心系统.05．中心计时器")
local addDelayedCallback = ____require_result_0.addDelayedCallback
local getServerTime = ____require_result_0.getServerTime
local GetHandleId = jass.GetHandleId
local IsUnitType = jass.IsUnitType
local UnitAddType = jass.UnitAddType
local UnitRemoveType = jass.UnitRemoveType
local UNIT_TYPE_SAPPER = jass.UNIT_TYPE_SAPPER
local _____8D1F_9762_6548_679C_514D_75AB_8868 = {}
local function _____53D6_5355_4F4D_53E5_67C4(_____5355_4F4D)
    if _____5355_4F4D == nil or _____5355_4F4D == 0 then
        return 0
    end
    return GetHandleId(_____5355_4F4D) or 0
end
local function _____7C7B_578B_5339_914D(typeName, typePrefix)
    if type(typeName) ~= "string" or typeName == "" or typePrefix == "" then
        return false
    end
    return __TS__StringSubstring(typeName, 0, #typePrefix) == typePrefix
end
local function _____8303_56F4_5305_542B(_____914D_7F6E_8303_56F4, _____76EE_6807_8303_56F4)
    if _____914D_7F6E_8303_56F4 == nil then
        return _____76EE_6807_8303_56F4 == "全部负面"
    end
    if type(_____914D_7F6E_8303_56F4) == "string" then
        return _____914D_7F6E_8303_56F4 == _____76EE_6807_8303_56F4
    end
    do
        local i = 0
        while i < #_____914D_7F6E_8303_56F4 do
            if _____914D_7F6E_8303_56F4[i + 1] == _____76EE_6807_8303_56F4 then
                return true
            end
            i = i + 1
        end
    end
    return false
end
local function _____53D6_6216_5EFA_8D1F_9762_6548_679C_514D_75AB_884C(_____5355_4F4D)
    local hid = _____53D6_5355_4F4D_53E5_67C4(_____5355_4F4D)
    if hid == 0 then
        return nil
    end
    local row = _____8D1F_9762_6548_679C_514D_75AB_8868[hid]
    if row == nil then
        row = {
            ["单位"] = _____5355_4F4D,
            ["原本是自爆工兵"] = IsUnitType(_____5355_4F4D, UNIT_TYPE_SAPPER) == true,
            ["全部负面免疫到期"] = 0,
            ["控制负面免疫到期"] = 0,
            ["魔法负面免疫到期"] = 0
        }
        _____8D1F_9762_6548_679C_514D_75AB_8868[hid] = row
    else
        row["单位"] = _____5355_4F4D
    end
    return row
end
local function _____884C_662F_5426_4ECD_6709_8D1F_9762_6548_679C_514D_75AB(row, now)
    return row["全部负面免疫到期"] > now or row["控制负面免疫到期"] > now or row["魔法负面免疫到期"] > now
end
local function _____5C1D_8BD5_6E05_7406_8D1F_9762_6548_679C_514D_75AB_884C(_____5355_4F4D)
    local hid = _____53D6_5355_4F4D_53E5_67C4(_____5355_4F4D)
    if hid == 0 then
        return
    end
    local row = _____8D1F_9762_6548_679C_514D_75AB_8868[hid]
    if row == nil then
        return
    end
    local now = getServerTime()
    if _____884C_662F_5426_4ECD_6709_8D1F_9762_6548_679C_514D_75AB(row, now) then
        return
    end
    if not row["原本是自爆工兵"] then
        UnitRemoveType(row["单位"], UNIT_TYPE_SAPPER)
    end
    __TS__Delete(_____8D1F_9762_6548_679C_514D_75AB_8868, hid)
end
local function ____on_8D1F_9762_6548_679C_514D_75AB_5230_671F()
    local now = getServerTime()
    for key in pairs(_____8D1F_9762_6548_679C_514D_75AB_8868) do
        local hid = key
        local row = _____8D1F_9762_6548_679C_514D_75AB_8868[hid]
        if row ~= nil and not _____884C_662F_5426_4ECD_6709_8D1F_9762_6548_679C_514D_75AB(row, now) then
            if not row["原本是自爆工兵"] then
                UnitRemoveType(row["单位"], UNIT_TYPE_SAPPER)
            end
            __TS__Delete(_____8D1F_9762_6548_679C_514D_75AB_8868, hid)
        end
    end
end
____exports["施加单位负面效果免疫"] = function(_____5355_4F4D, _____6301_7EED_65F6_95F4, _____53C2_6570)
    if _____53C2_6570 == nil then
        _____53C2_6570 = {}
    end
    if _____5355_4F4D == nil or _____5355_4F4D == 0 or not (_____6301_7EED_65F6_95F4 > 0) then
        return
    end
    local row = _____53D6_6216_5EFA_8D1F_9762_6548_679C_514D_75AB_884C(_____5355_4F4D)
    if row == nil then
        return
    end
    local now = getServerTime()
    local ____until = now + _____6301_7EED_65F6_95F4 * 1000
    local _____8303_56F4 = _____53C2_6570["范围"]
    if _____8303_56F4_5305_542B(_____8303_56F4, "全部负面") then
        row["全部负面免疫到期"] = row["全部负面免疫到期"] > ____until and row["全部负面免疫到期"] or ____until
    end
    if _____8303_56F4_5305_542B(_____8303_56F4, "控制") then
        row["控制负面免疫到期"] = row["控制负面免疫到期"] > ____until and row["控制负面免疫到期"] or ____until
    end
    if _____8303_56F4_5305_542B(_____8303_56F4, "魔法负面") then
        row["魔法负面免疫到期"] = row["魔法负面免疫到期"] > ____until and row["魔法负面免疫到期"] or ____until
    end
    if _____53C2_6570["同步原生技能目标免疫"] ~= false then
        UnitAddType(_____5355_4F4D, UNIT_TYPE_SAPPER)
    end
    addDelayedCallback(_____6301_7EED_65F6_95F4 * 1000 + 50, ____on_8D1F_9762_6548_679C_514D_75AB_5230_671F)
end
____exports["施加单位控制负面效果免疫"] = function(_____5355_4F4D, _____6301_7EED_65F6_95F4, _____540C_6B65_539F_751F_6280_80FD_76EE_6807_514D_75AB)
    if _____540C_6B65_539F_751F_6280_80FD_76EE_6807_514D_75AB == nil then
        _____540C_6B65_539F_751F_6280_80FD_76EE_6807_514D_75AB = true
    end
    ____exports["施加单位负面效果免疫"](_____5355_4F4D, _____6301_7EED_65F6_95F4, {["范围"] = "控制", ["同步原生技能目标免疫"] = _____540C_6B65_539F_751F_6280_80FD_76EE_6807_514D_75AB})
end
____exports["施加单位魔法负面效果免疫"] = function(_____5355_4F4D, _____6301_7EED_65F6_95F4, _____540C_6B65_539F_751F_6280_80FD_76EE_6807_514D_75AB)
    if _____540C_6B65_539F_751F_6280_80FD_76EE_6807_514D_75AB == nil then
        _____540C_6B65_539F_751F_6280_80FD_76EE_6807_514D_75AB = true
    end
    ____exports["施加单位负面效果免疫"](_____5355_4F4D, _____6301_7EED_65F6_95F4, {["范围"] = "魔法负面", ["同步原生技能目标免疫"] = _____540C_6B65_539F_751F_6280_80FD_76EE_6807_514D_75AB})
end
____exports["清除单位负面效果免疫"] = function(_____5355_4F4D)
    local hid = _____53D6_5355_4F4D_53E5_67C4(_____5355_4F4D)
    if hid == 0 then
        return
    end
    local row = _____8D1F_9762_6548_679C_514D_75AB_8868[hid]
    if row == nil then
        return
    end
    row["全部负面免疫到期"] = 0
    row["控制负面免疫到期"] = 0
    row["魔法负面免疫到期"] = 0
    _____5C1D_8BD5_6E05_7406_8D1F_9762_6548_679C_514D_75AB_884C(_____5355_4F4D)
end
____exports["单位是否免疫负面效果类型"] = function(_____5355_4F4D, typeName)
    local hid = _____53D6_5355_4F4D_53E5_67C4(_____5355_4F4D)
    if hid == 0 then
        return false
    end
    local row = _____8D1F_9762_6548_679C_514D_75AB_8868[hid]
    if row == nil then
        return false
    end
    local now = getServerTime()
    if not _____884C_662F_5426_4ECD_6709_8D1F_9762_6548_679C_514D_75AB(row, now) then
        _____5C1D_8BD5_6E05_7406_8D1F_9762_6548_679C_514D_75AB_884C(_____5355_4F4D)
        return false
    end
    if row["全部负面免疫到期"] > now and _____7C7B_578B_5339_914D(typeName, "Debuff:") then
        return true
    end
    if row["控制负面免疫到期"] > now and _____7C7B_578B_5339_914D(typeName, "Debuff:control") then
        return true
    end
    if row["魔法负面免疫到期"] > now and _____7C7B_578B_5339_914D(typeName, "Debuff:magic") then
        return true
    end
    return false
end
____exports["单位是否免疫负面效果BuffID"] = function(_____5355_4F4D, buffID)
    local meta = buffTableMod.buffs[buffID]
    return ____exports["单位是否免疫负面效果类型"](_____5355_4F4D, meta and meta.type)
end
return ____exports
